Point Estimate of Mean

In [1]:
import numpy as np

heights = [65.2, 67.1, 68.5, 69.3, 70.0,
           71.2, 72.4, 73.1, 74.5, 75.8]

point_estimate = np.mean(heights)
print("Point Estimate (Mean):", point_estimate)


Point Estimate (Mean): 70.71


Interval Estimate Using StatsModels – DescrStatsW

In [4]:
from statsmodels.stats.weightstats import DescrStatsW
import numpy as np

data = np.array(heights)

desc = DescrStatsW(data)

print("Point Estimate (Mean):", desc.mean)
print("95% Confidence Interval:", desc.tconfint_mean())


Point Estimate (Mean): 70.71000000000001
95% Confidence Interval: (np.float64(68.3272779750514), np.float64(73.09272202494861))


Normally, to find a confidence interval you must:

Compute mean

Compute standard deviation

Compute standard error

Find t-critical value

Compute margin of error

Build confidence interval

👉 DescrStatsW does all of this internally for you.

In [2]:
import numpy as np
from scipy import stats

heights = np.array([65.2, 67.1, 68.5, 69.3, 70.0,
                    71.2, 72.4, 73.1, 74.5, 75.8])

sigma = 3      # known population std
n = len(heights)
mean = np.mean(heights)

ci = stats.norm.interval(
    0.95,
    loc=mean,
    scale=sigma / np.sqrt(n)
)

print("Point Estimate:", mean)
print("95% CI (σ known):", ci)


Point Estimate: 70.71
95% CI (σ known): (np.float64(68.85061490308631), np.float64(72.56938509691368))


| Scenario                | Correct Tool            |
| ----------------------- | ----------------------- |
| σ unknown, small sample | `DescrStatsW`           |
| σ unknown, general      | `stats.t.interval()`    |
| σ known                 | `stats.norm.interval()` |
| Hypothesis testing      | `scipy.stats` tests     |


Implementing P-value in Python

Let's consider a scenario where a tutor believes that the average exam score of their students is equal to the national average (85). The tutor collects a sample of exam scores from their students and performs a one-sample t-test to compare it to the population mean (85).

The code performs a one-sample t-test to compare the mean of a sample data set to a hypothesized population mean.
It utilizes the scipy.stats library to calculate the t-statistic and p-value. SciPy is a Python library that provides efficient numerical routines for scientific computing.
The p-value is compared to a significance level (alpha) to determine whether to reject the null hypothesis.

Null Hypothesis (H₀)

H0​:μ=85

Alternative Hypothesis (Hₐ)

Ha​:μ!=85

In [3]:
import scipy.stats as stats

# exam scores
sample_data = [78, 82, 88, 95, 79, 92, 85, 88, 75, 80]

# Population mean
population_mean = 85

# One-sample t-test
t_stat, p_value = stats.ttest_1samp(sample_data, population_mean)

print("t-statistic:", t_stat)
print("p-value:", p_value)

# Conditions
alpha = 0.05
if p_value < alpha:
    print("Reject the null hypothesis. There is enough evidence to suggest a significant difference.")
else:
    print("Fail to reject the null hypothesis. The difference is not statistically significant.")

t-statistic: -0.3895364838967159
p-value: 0.7059365203154573
Fail to reject the null hypothesis. The difference is not statistically significant.


**Real life Examples of Hypothesis Testing**

A pharmaceutical company tests a new drug to see if it lowers blood pressure in patients.

Data:

Before Treatment: 120, 122, 118, 130, 125, 128, 115, 121, 123, 119
After Treatment: 115, 120, 112, 128, 122, 125, 110, 117, 119, 114

Step 1: Define the Hypothesis

Null Hypothesis: (H0)The new drug has no effect on blood pressure.
Alternate Hypothesis: (H1)The new drug has an effect on blood pressure.

In [4]:
from scipy import stats
import numpy as np

# Blood pressure readings
before = np.array([120, 122, 118, 130, 125, 128, 115, 121, 123, 119])
after  = np.array([115, 120, 112, 128, 122, 125, 110, 117, 119, 114])

# Paired t-test
# alternative='greater' because we expect before > after
t_stat, p_value = stats.ttest_rel(before, after, alternative='greater')

print("t-statistic:", t_stat)
print("p-value:", p_value)

# Significance level
alpha = 0.05

if p_value < alpha:
    print("Reject H0: Drug has a significant effect on lowering blood pressure")
else:
    print("Fail to reject H0: No significant effect observed")


t-statistic: 9.0
p-value: 4.269025611583142e-06
Reject H0: Drug has a significant effect on lowering blood pressure


A small research group is studying the weights of a rare breed of miniature dogs. They collect the weights (in pounds) from a pilot study sample.

The recorded sample weights are:


35,42,48,55,40,39,41,52,45,43

In [2]:
from scipy.stats import sem
import numpy as np

# Sample data (e.g., the dog weights)
data = np.array([35, 42, 48, 55, 40, 39, 41, 52, 45, 43])

# Calculate the Standard Error of the Mean
standard_error = sem(data)

print(f"Calculated Standard Error ): {standard_error:.4f}")

Calculated Standard Error ): 1.9379


In [3]:
from scipy.stats import t
import numpy as np

# Sample statistics (reuse previous results)
mean = np.mean(data)
n = len(data)
df = n - 1

# 95% Confidence Interval
ci_lower, ci_upper = t.interval(
    confidence=0.95,
    df=df,
    loc=mean,
    scale=standard_error
)

print(f"95% Confidence Interval: ({ci_lower:.2f}, {ci_upper:.2f})")


95% Confidence Interval: (39.62, 48.38)


We are 95% confident that the true mean weight of all dogs of this rare miniature breed lies between 39.62 pounds and 48.38 pounds.